src/libtools/calculator.cc

```diff
@@ -75,5 +75,66 @@ void Calculator::OverwriteDefaultsWithUserInput(const Property &p,
   }
 }
 
+std::vector<std::string> Calculator::GetPropertyChoices(const Property &p) {
+  if (p.hasAttribute("choices")) {
+    std::string att = p.getAttribute<std::string>("choices");
+    Tokenizer tok{att, " ,"};
+    return tok.ToVector();
+  } else {
+    return {""};
+  }
+}
+
+void Calculator::RecursivelyCheckOptions(const Property &p) {
+  for (const Property &prop : p) {
+    if (prop.HasChildren()) {
+      RecursivelyCheckOptions(prop);
+    } else {
+      std::vector<std::string> choices = Calculator::GetPropertyChoices(prop);
+      const std::string &head = choices.front();
+      if (head != "") {
+        try {
+          Calculator::CheckOption(prop, choices);
+        } catch (const std::runtime_error &e) {
+          std::ostringstream oss;
+          oss << "\nInput option: " << prop.name() << "\n";
+          oss << "Must be one of: ";
+          for (const std::string c : choices) {
+            oss << c << "\n";
+          }
+          std::cout << oss.str();
+          throw e;
+        }
+      }
+    }
+  }
+}
+
+void Calculator::CheckOption(const Property &prop,
+                             const std::vector<std::string> &choices) {
+  const std::string &head = choices.front();
+  if (head == "bool") {
```

Comment:
what happens if the input cannot be converted to bool?

Reply:
If it cannot convert the input to `bool`, `float` and `int` it raises an error. I am going to add a test for it :)

Reply:
yeah but the error is not very descriptive is it? Btw. it is weekend. :)